srcCpp03/RTIDDSImpl.cxx

```diff
@@ -564,6 +589,110 @@ class RTIPublisher: public RTIPublisherBase<T> {
+
+          for (int j = 0; j < KEY_SIZE; ++j) {
+              // The key will be i but splitted in bytes
+              key_offset.set_element(j, (unsigned char) (i >> j * 8));
+          }
+      }
+
+  public:
+      RTIFlatDataPublisher(
+              dds::pub::DataWriter<T> writer,
+              int num_instances,
+              rti::core::Semaphore& pongSemaphore,
+              bool useSemaphore,
+              int instancesToBeWritten,
+              ParameterManager *PM
+              )
+              :
+              RTIPublisherBase<T>(
+                      writer,
+                      num_instances,
+                      pongSemaphore,
+                      useSemaphore,
+                      instancesToBeWritten,
+                      PM)
+      {
+          for (unsigned long int i = 0; i < this->_num_instances; ++i) {
+              Builder builder = rti::flat::build_data(writer);
+              add_key(builder, i);
+
+              T *sample = builder.finish_sample();
+
+              this->_instance_handles.push_back(
+                  this->_writer.register_instance(*sample));
+
+              this->_writer.extensions().discard_loan(*sample);
+          }
+
+          // Register the key of MAX_CFT_VALUE
+          Builder builder = rti::flat::build_data(writer);
+          add_key(builder, MAX_CFT_VALUE);
+          T *sample = builder.finish_sample();
+
+          this->_instance_handles.push_back(
+              this->_writer.register_instance(*sample));
+
+          this->_writer.extensions().discard_loan(*sample);
+      }
+
+      inline bool send(TestMessage &message, bool isCftWildcardKey) {
+          Builder builder = rti::flat::build_data(this->_writer);
+          long key = 0;
+
+          // Initialize Information data
+          builder.add_entity_id(message.entity_id);
+          builder.add_seq_num(message.seq_num);
+          builder.add_timestamp_sec(message.timestamp_sec);
+          builder.add_timestamp_usec(message.timestamp_usec);
+          builder.add_latency_ping(message.latency_ping);
+
+          // Add payload
+          BinDataBuilder bin_data_builder = builder.build_bin_data();
+          bin_data_builder.add_n(message.size);
+          bin_data_builder.finish();
+
+          // calculate key and add it
+          if (!isCftWildcardKey && this->_num_instances > 1) {
+              key = (this->_instancesToBeWritten == -1)
+                      ? this->_instance_counter++ % this->_num_instances
+                      : this->_instancesToBeWritten;
+          } else {
+              key = MAX_CFT_VALUE;
+          }
+
+          add_key(builder, key);
+
+          // Build the data to be sent
+          T *sample = builder.finish_sample();
+
+          // Send data through the writer
+          if (!isCftWildcardKey) {
+              // .back() because the CFT Handler is always the last one to be pushed back
+              this->_writer.write(*sample, this->_instance_handles.back());
```

Comment:
Minor enhancement: instead of writing a comment to indicate this invariant, make it explicit in code. For example, define a private member function `get_cft_instance_handle() { return _instance_handles.back(); }`

Don't worry about performance; the compiler will inline it.

Reply:
Ok, good idea!